Document RequestStreamHandler, JUnit tests, Gradle, AWS SDK Java v2, and AWS ASK SDK

oztimpower opened the pull request:
Enhancement to the Amazon Lambda extension documentation, primarily driven by the addition of the RequestSteamHandler.

I have also added some of the more difficult topics around AWS Lambda with native image, such as using the AWS Java SDK v2 SSL client, and Amazon Alexa with the ASK SDK v2.

Gradle support was added also for those looking for a simpler view on how to use Gradle with the Amazon Lambda extension (as opposed to having to stitch it all together themselves).  Indeed that's the theme with all these additions, I've spent months fiddling with all this stuff in Java & Kotlin, Gradle and Maven, JVM and Native Image, trying to get it all to work.  I do plan to write a few blogs, however I wanted to capture as much of those learnings into the Quarkus docs directly for better re-use.

Yes, some of the additions for AWS Java SDK v2 and Alexa ASK SDK v2, could be a guide on their own, as there is a lot more to add.  I plan to write my own Quarkus extensions to automate all of this, so arguably the documentation will reduce in the future, and perhaps could remain in one place.  

/cc @patriot1burke 
/cc @evanchooly 
/cc @bnusunny

oztimpower committed:
Include Gradle support, JUnit tests, AWS SDK Java v2

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+    "methods":[{"name":"<init>","parameterTypes":[] }]
+  }
+]
+----
```
I think it would probably be better to use the `quarkus-jaxb` extension. I think it adds that class for reflection.

oztimpower replied:
Thanks for the tip.  I confirm that works in my sample project, updating the docs accordingly.

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+  client = SqsClient.builder().region(Region.XXXX).httpClient(software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient.builder().build()).build();
+----
+
+For Maven, `add` the following to your `pom.xml`.
```
```suggestion
For Maven, add the following to your `pom.xml`.
```

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+
+[source]
+----
+docker cp {container-id-from-above}:/opt/graalvm/jre/lib/amd64/libsunec.so src/main/resources/
```
This is not needed anymore with GraalVM 19.3. As it's the minimal version now, we can get rid if that sunec thing.

oztimpower replied:
Unfortunately that's not the case, I wish it was.  GraalVM 19.3.1 or otherwise.  I've tried Quarkus 1.2.0.Final, 1.2.1.Final, 1.3.0.Final and 999-SNAPSHOT (HEAD) and it's all the same.  

For example see #7846.

For anyone going through that pain (which can turn them away from Quarkus & native-image), I think there's value in having the info there, until it never happens again.  Happy to also publish a reproducible project to show this, it's how I've generated the detail for the documentation featured here.

With 1.3.0.Final and GraalVM 19.3.1 (and native-image SSL enabled):


```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Running Quarkus native-image plugin on GraalVM Version 19.3.1 CE

286 | [INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] /opt/graalvm/bin/native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Dcom.sun.xml.bind.v2.bytecode.ClassTailor.noOptimize=true -J-Duser.language=en -J-Dfile.encoding=ANSI_X3.4-1968 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -H:+JNI -jar gw-sqs-push-1.0-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:-IncludeAllTimeZones -H:EnableURLProtocols=http,https --enable-all-security-services -H:NativeLinkerOption=-no-pie --no-server -H:-UseServiceLoaderFeature -H:+StackTrace gw-sqs-push-1.0-SNAPSHOT-runner

```


Invoking function (sh manage.sh native invoke)

```
483 | ++ aws lambda invoke response.txt --function-name GwSqsPushNativeFunction --payload file://payload.json --log-type Tail --query LogResult --output text
484 | ++ base64 -d
485 | START RequestId: ea42071e-111e-4d19-b1ee-3da17e54d7cf Version: $LATEST
486 | handleRequest: Invoked with request [ SQSRequest(firstName=Timothy, lastName=Power) ]
487 | 2020-03-19 05:25:42,649 INFO  [org.tjp.LambdaSQSPush] (Lambda Thread) handleRequest: Invoked with request [ SQSRequest(firstName=Timothy, lastName=Power) ]
488 |  
489 | handleRequest: Error [ Unable to execute HTTP request: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty ]
490 | [Ljava.lang.StackTraceElement;@7fd3712e4c98END RequestId: ea42071e-111e-4d19-b1ee-3da17e54d7cf
491 | REPORT RequestId: ea42071e-111e-4d19-b1ee-3da17e54d7cf  Duration: 428.63 ms Billed Duration: 500 ms Memory Size: 256 MB Max Memory Used: 40 MB  Init Duration: 18.13 ms
492 | {"body":"{ \"exception\": \"Unable to execute HTTP request: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty\" }","headers":{"Content-Type":"application/json","X-Custom-Header":"application/json"},"statusCode":500}/codebuild/output/src105077389/src

```

oztimpower replied:
Marking this as resolved, since it's still an issue, is that ok?

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+    -Djavax.net.ssl.trustStorePassword=changeit
+----
+
+You final archive will look like this:
```
```suggestion
Your final archive will look like this:
```

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+    ssl/libsunec.so
+----
+
+This can be automated with a new Maven assembly zip.xml, for a new profile, such function-zip.xml
```
```suggestion
This can be automated with a new Maven assembly zip.xml, for a new profile, such as the `function-zip.xml` below:
```

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+        </file>
+        <file>
+            <source>${project.basedir}${file.separator}src${file.separator}main${file.separator}resources${file.separator}libsunec.so</source>
+            <outputDirectory>${file.separator}ssl</outputDirectory>
```
As mentioned above, this shouldn't be needed anymore.

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+== Writing Skills with the Amazon Alexa ASK SDK v2
+
+The starting point for an Alexa Skill on AWS Lambda is typically via subclassing `SkillStreamHandler`,
+which implements the RequestStreamHandler::requestHandler.  If you examine the source code for this class it is very simple,
```
```suggestion
which implements the `RequestStreamHandler::requestHandler` method.  If you examine the source code for this class it is very simple,
```

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+which implements the RequestStreamHandler::requestHandler.  If you examine the source code for this class it is very simple,
+and can be implemented directly without this helper base class.
+
+NOTE: Quarkus requires a direct implementation of the RequestStreamHandler, as the build process will scan only directly implemented interfaces,
```
```suggestion
NOTE: Quarkus requires a direct implementation of `RequestStreamHandler`, as the build process will scan only directly implemented interfaces,
```

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+and can be implemented directly without this helper base class.
+
+NOTE: Quarkus requires a direct implementation of the RequestStreamHandler, as the build process will scan only directly implemented interfaces,
+inherited interfaces will not be seen.
```
Why don't we fix that instead?

oztimpower replied:
Yes... part of the reason I added this to the docs was to draw attention to this.  If a core member (or someone who understands what to change in the core libraries) could help make the change that would be appreciated.  I'll raise a separate Feature Request item for it.

oztimpower replied:
Followed the instruction on #7955 to check for subclasses of the SkillStreamHandler also.  I have this working (2 lines of code, plus runtime bom addition).

Hence, I'm removing all Alexa references from the document.

gsmet commented on docs/src/main/asciidoc/amazon-lambda.adoc:
```diff
@@ -240,33 +294,469 @@ of things you need to do.  Take a look at the generated example project to get a
+NOTE: Quarkus requires a direct implementation of the RequestStreamHandler, as the build process will scan only directly implemented interfaces,
+inherited interfaces will not be seen.
+
+Specifically you add a requestHandler implementation as below.
```
```suggestion
Specifically you add a `RequestStreamHandler` implementation as below.
```

gsmet reviewed (changes requested):
Thanks for this work! It's awesome.

I added a bunch of minor suggestions, mostly typos and formatting but also asked a couple of questions.

Could you have a look?

patriot1burke commented:
@oztimpower Really nice PR!  I'm freeing up some next week and will work to help resolve all the issues you and others have uncovered.

oztimpower commented:
@gsmet Ok, I've made all the changes (typos et al), and removed the Alexa references per your request to fix it directly in the Amazon Lambda Deployment (PR coming).

The quarkus-jaxb extension did fix the need for a reflection-config.json, so that is removed from the documentation, and updated.

I've left in the libsunec.so stuff as that is still happening for me and others on GraalVM 19.3.1

Can merge please?

gsmet commented:
> I've left in the libsunec.so stuff as that is still happening for me and others on GraalVM 19.3.1

This is a problem. We need to check that as we removed it from other documentation.

@gwenneg any chance you could check that again?
@oztimpower totally unrelated to your PR so I'll check it later today and hopefully merge it.

gsmet commented:
Nice work @oztimpower !
